src/cupbearer/detectors/statistical/mahalanobis_detector.py

```diff
@@ -6,11 +6,12 @@
 )
 
 
-def _pinv(C, rcond):
+def _pinv(C, rcond, dtype=torch.float64):
     # Workaround for pinv not being supported on MPS
+    og_dtype = C.dtype
     if C.is_mps:
-        return torch.linalg.pinv(C.cpu(), rcond=rcond, hermitian=True).to(C.device)
-    return torch.linalg.pinv(C, rcond=rcond, hermitian=True)
+        return torch.linalg.pinv(C.cpu().to(dtype), rcond=rcond, hermitian=True).to(og_dtype).to(C.device)
+    return torch.linalg.pinv(C.to(dtype), rcond=rcond, hermitian=True)
```

Comment:
```suggestion
    return torch.linalg.pinv(C.to(dtype), rcond=rcond, hermitian=True).to(C.dtype)
```
Return result should preserve dtype here as well. If you want to keep `og_dtype`, use that here.